shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -423,12 +426,15 @@ - (void)viewDidDisappear:(BOOL)animated {
   TRACE_EVENT0("flutter", "viewDidDisappear");
   [self surfaceUpdated:NO];
   [[_engine.get() lifecycleChannel] sendMessage:@"AppLifecycleState.paused"];
-
+  [_engine.get() setViewController:nil];
   [super viewDidDisappear:animated];
 }
 
 - (void)dealloc {
   [[NSNotificationCenter defaultCenter] removeObserver:self];
+  if (_engineIsOwnedByMe) {
```

Comment:
I haven't fully digested this, but my initial impression is that this feels clunky.  My instinct is that we shouldn't need to keep track of this explicitly with proper ref-counting and strong/weak references.

That is, could `FlutterViewController` always own *its* reference to the `FlutterEngine`, everything else that doesn't need to keep the engine alive could use weak references, and anybody that does want to keep the engine alive could be responsible for retaining their own reference to the `FlutterEngine`?

(Is it important that we be able to shut down the engine without deallocating it?)

Reply:
The problem is that we end up with a cyclic reference in the case where `FlutterViewController` initializes its own engine. Basically, it creates the engine (and it needs a strong reference to the engine), and then tells the engine to use it (weakly).  When some external action that dismisses the view controller happens, it doesn't currently notify the engine that it can be dismissed - and so it never gets deallocated.  

__Just__ notifying the engine that it's not visible isn't enough either, because the engine currently assumes that once it's started it should never stop unless deallocated - but we end up with a circle of references that prevents that from happening.  Basically, we should be treating a deallocating of the view controller as something that needs to shut down the engine IFF the engine is owned by the view controller.

Reply:
In other words, a `FlutterEngine` is built such that it will keep running whether it has a `FlutterViewController` or not.  But in this case, we want to tell it to stop running once the `FlutterViewController` is gone.  

Stated that way, I'll try to see if I can alternatively add a parameter or initializer to `FlutterEngine` that would ask whether it should be allowed to run headless or not.